src/Microsoft.Dnx.Watcher.Core/ProjectWatcher.cs

```diff
@@ -0,0 +1,286 @@
+using System.IO;
+using System.Linq;
+using System.Runtime.Versioning;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Dnx.Compilation;
+using Microsoft.Dnx.Runtime;
+using Microsoft.Dnx.Runtime.Common.Impl;
+using Microsoft.Framework.Logging;
+
+namespace Microsoft.Dnx.Watcher.Core
+{
+    public class ProjectWatcher
+    {
+        private readonly Func<string, IFileWatcher> _fileWatcherFactory;
+        private readonly IProjectGraphProvider _projectGraphProvider;
+
+        private readonly ILoggerFactory _loggerFactory;
+        private readonly ILogger _logger;
+
+        private string _fullProjectFilePath;
+        private string _rootPath;
+
+        public ProjectWatcher(
+            Func<string, IFileWatcher> fileWatcherFactory,
+            IProjectGraphProvider projectGraphProvider,
+            ILoggerFactory loggerFactory)
+        {
+            _fileWatcherFactory = fileWatcherFactory;
+            _projectGraphProvider = projectGraphProvider;
+            _loggerFactory = loggerFactory;
+            _logger = _loggerFactory.CreateLogger(nameof(ProjectWatcher));
+        }
+
+        private string ResolveProjectFileToWatch(string projectOrDirectory)
+        {
+            if (string.IsNullOrEmpty(projectOrDirectory))
+            {
+                projectOrDirectory = Directory.GetCurrentDirectory();
+            }
+
+            if (!string.Equals(Path.GetFileName(projectOrDirectory), Project.ProjectFileName, StringComparison.Ordinal))
+            {
+                projectOrDirectory = Path.Combine(projectOrDirectory, Project.ProjectFileName);
+            }
+
+            if (!File.Exists(projectOrDirectory))
+            {
+                // TODO: better error message
+                throw new InvalidOperationException("Project file not found");
+            }
+
+            return projectOrDirectory;
+        }
+
+        public Task<bool> WatchAsync(string projectOrDirectory, string[] dnxArguments, string workingDir, CancellationToken cancellationToken)
+        {
+            _fullProjectFilePath = ResolveProjectFileToWatch(projectOrDirectory);
+            _rootPath = Path.GetDirectoryName(_fullProjectFilePath);
+
+            dnxArguments = new string[] { "--project", _fullProjectFilePath }
+               .Concat(dnxArguments)
+               .Select(arg =>
+               {
+                    // If the argument has spaces, make sure we quote it
+                    if (arg.Contains(" ") || arg.Contains("\t"))
+                   {
+                       return $"\"{arg}\"";
+                   }
+
+                   return arg;
+               })
+               .ToArray();
+
+            return WatchAsync(dnxArguments, workingDir, cancellationToken);
+        }
+
+        private async Task<bool> WatchAsync(string[] dnxArguments, string workingDir, CancellationToken cancellationToken)
+        {
+            var dnxArgumentsAsString = string.Join(" ", dnxArguments);
```

Comment:
Unfortunately what the script executor does is not helpful here because I need to convert an array back to string arguments. There's no place where we do this where the quotes have been already removed